File: algorithms/cpp/287.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution
{
public:
    int findDuplicate(vector<int>& nums)
    {
        int slow = nums.size()-1, fast = slow;
        while ( true )
        {
            slow = nums[slow]-1;
            fast = nums[nums[fast]-1]-1;
            if ( slow == fast )
                break;
        }
        fast = nums.size()-1;
        while ( true )
        {
            slow = nums[slow]-1;
            fast = nums[fast]-1;
            if ( slow == fast )
                return slow+1;
        }
    }
};

int main()
{
    return 0;
}

File: algorithms/rust/287.rs
impl Solution {
    #[allow(dead_code)]
    pub fn find_duplicate(mut nums: Vec<i32>) -> i32 {
        let n = nums.len() - 1;
        for i in 1..=n {
            while nums[i] != i as i32 {
                let t = nums[i] as usize;
                if nums[i] == nums[t] {
                    return nums[i];
                }
                nums[i] = nums[t];
                nums[t] = t as i32;
            }
        }
        nums[0]
    }
}

struct Solution;

fn main() {
    println!("Hello, world!");
}
